### ISSUE TITLE
Document picocli use with java modules

### ISSUE THREAD

**Issue by remkop** [remkop/picocli#291]

**Comment by remkop** [comment]
Leave a comment

**Comment by remkop** [comment]
@kaspernielsen Can you explain a bit more about what you’re looking to do and how picocli support for `MethodHandles.Lookup` would help?

**Comment by kaspernielsen** [comment]
@remkop Here is a good overview of MethodHandles.Lookup vs explicitly opening the packages.
http://in.relation.to/2017/04/11/accessing-private-state-of-java-9-modules/

Basically, I would be able to just passe a MethodHandles.Lookup to Picocli instead of having to modify module-info.java.

**Comment by kaspernielsen** [comment]
@remkop No, you would only need to pass it once per module. It captures the permissions of the module that creates the MethodHandles.Lookup object.

Right now GraalVM does not support MethodHandles. There are two types the Direct ones which are basically like normal method, you can actually wrap and unwrap to normal methods. I'm pretty sure they will be supported once when they release GraalVM for Java 11. Non-direct ones where you can build "code" at runtime will most likely not be supported. 

Sounds cool with annotation processor, hope it will still be possible to use to do it runtime as well.

 and

**Comment by kaspernielsen** [comment]
@remkop Just to follow up on this. If you want to access fields on classes that are located in named modules. You basically have 2 choices going forward, module-info.java or MethodHandles.Lookup. 
I can't see how you would be able to generate accessors without using one way of the two.
MethodHandles.Lookup definitely being the most flexible.

I also think you will have a hard time keeping both <Java 9 and >=Java 9 in the same codebase. At least if you want it as compact as now.

**Comment by remkop** [comment]
This may not be as difficult as I thought. Picocli already has an abstraction mechanism for getting and setting values from annotated fields and methods: the `CommandLine.Model.IGetter` and `CommandLine.Model.ISetter` interfaces.

Providing a  `MethodHandles.Lookup`-based implementation of these interfaces should not be too difficult. The problem is wiring things up, since currently during the construction of a `CommandLine` object, picocli will initialize these getters and setters with reflection-based implementations. If the client application's `module-info` does not open the application to picocli, an `InaccessibleObjectException` is thrown during the construction of a `CommandLine` object.

One idea for addressing this would be to introduce a new interface `IPrivateAccessStrategy`. There would need to be additional `CommandLine(Object, IFactory, IPrivateAccessStrategy)` constructors and `CommandSpec.forAnnotatedObject(Object, IFactory, IPrivateAccessStrategy)` factory methods to allow clients to specify an instance.

The existing `CommandLine` constructors would use a default `IPrivateAccessStrategy` implementation that creates the current reflection-based field/method getters and setters.

A custom `IPrivateAccessStrategy` implementation could create `MethodHandles.Lookup`-based getters and setters. This implementation could live in the `picocli-jpms-module` subproject. Java 9-based client applications could use code like below to construct a `CommandLine` object:

```java
import picocli.CommandLine.Model.IPrivateAccessStrategy;
import picocli.module.LookupAccessStrategy; // in picocli-jpms-module

@Command
class MyApp {
    public static void main(String... args) {
        MethodHandles.Lookup lookup = MethodHandles.lookup(); // client code creates the Lookup
        IPrivateAccessStrategy privateAccess = new LookupAccessStrategy(lookup);
        CommandLine cmd = new CommandLine(new MyApp(), privateAccess);
        cmd.execute(args);
    }
    // ...
}
```

The contract of this interface could look something like this:

```java
interface IPrivateAccessStrategy {
    IGetter createFieldValueGetter(Field, IScope, CommandSpec);
    ISetter createFieldValueSetter(Field, IScope, CommandSpec);
    IGetter createMethodValueGetter(Method, IScope, CommandSpec);
    ISetter createMethodValueSetter(Method, IScope, CommandSpec);
}
```

**Comment by kaspernielsen** [comment]
That sounds very promising, I hope I will have some time to look into it this week. Thanks
